ingest/src/main/java/com/microsoft/azure/kusto/ingest/IngestClientBase.java

```diff
@@ -13,19 +17,33 @@
     }
 
     static String getIngestionEndpoint(String clusterUrl) {
-        if (clusterUrl.contains(INGEST_PREFIX)) {
+        if (clusterUrl == null || clusterUrl.contains(INGEST_PREFIX) || isReservedHostname(clusterUrl)) {
             return clusterUrl;
-        } else {
+        }
+        if (clusterUrl.contains(PROTOCOL_SUFFIX)) {
             return clusterUrl.replaceFirst(PROTOCOL_SUFFIX, PROTOCOL_SUFFIX + INGEST_PREFIX);
         }
+        return INGEST_PREFIX + clusterUrl;
     }
 
     static String getQueryEndpoint(String clusterUrl) {
-        if (clusterUrl.contains(INGEST_PREFIX)) {
-            return clusterUrl.replaceFirst(INGEST_PREFIX, "");
-        } else {
-            return clusterUrl;
-        }
+        return isReservedHostname(clusterUrl) ? clusterUrl : clusterUrl.replaceFirst(INGEST_PREFIX, "");
     }
 
+    static boolean isReservedHostname(String rawUri) {
+        URI uri = URI.create(rawUri);
+        if (!uri.isAbsolute()) {
+            return true;
+        }
+        var host = uri.getHost().toLowerCase();
+        Pattern IPV4_PATTERN = Pattern.compile("^(25[0-5]|2[0-4]\\d|[0-1]?\\d?\\d)(\\.(25[0-5]|2[0-4]\\d|[0-1]?\\d?\\d)){3}$");
```

Comment:
I told you not to import new libraries, but we already have com.apache.http.componenets, so you can use InetAddressUtils.isIPv6Address/InetAddressUtils.isIPv4Address